assembl/static2/js/app/main.jsx

```diff
@@ -1,58 +1,46 @@
 // @flow
-
 import * as React from 'react';
 import { connect } from 'react-redux';
 
-import { getCurrentPhaseIdentifier } from './utils/timeline';
+import { getCurrentPhaseData, getPhaseId } from './utils/timeline';
 import Navbar from './components/navbar/navbar';
 import Footer from './components/common/footer';
 import CookiesBar from './components/cookiesBar';
+import { fromGlobalId } from './utils/globalFunctions';
 
 type Props = {
   timeline: Timeline,
-  params: { phase: string, themeId: ?string },
-  location: { query: { phase: string }, pathname: string },
+  params: { phase?: string, phaseId?: string, themeId?: string },
+  location: { pathname: string },
   children: React.Node
 };
 
-type State = { identifier: string, location: string };
-
-class Main extends React.Component<Props, State> {
-  constructor(props: Props) {
-    super(props);
-    const { timeline } = this.props;
-    const paramsIdentifier = this.props.params.phase || getCurrentPhaseIdentifier(timeline);
-    const queryIdentifier = this.props.location.query.phase || paramsIdentifier;
-    this.state = {
-      identifier: queryIdentifier,
-      location: this.props.location.pathname
-    };
-  }
-
-  componentWillReceiveProps(nextProps) {
-    const location = nextProps.location.pathname;
-    const { timeline } = this.props;
-    const paramsIdentifier = nextProps.params.phase || getCurrentPhaseIdentifier(timeline);
-    const queryIdentifier = nextProps.location.query.phase || paramsIdentifier;
-    this.setState({
-      identifier: queryIdentifier,
-      location: location
-    });
-
-    window.scrollTo({ top: 0, left: 0, behavior: 'smooth' });
-  }
-
+class Main extends React.Component<Props> {
   render() {
-    const { themeId } = this.props.params;
-    const that = this;
+    const { params, timeline } = this.props;
+    const { themeId } = params;
+    const { currentPhaseIdentifier, currentPhaseId } = getCurrentPhaseData(timeline);
+    let identifier = params.phase || null;
+    let phaseId = params.phaseId || null;
+    if (!identifier) {
+      identifier = currentPhaseIdentifier;
+      phaseId = currentPhaseId;
+    }
+    if (!phaseId && identifier) {
+      // keep old shared urls working
+      phaseId = getPhaseId(timeline, identifier);
+    }
+    const discussionPhaseId = phaseId ? fromGlobalId(phaseId) : null;
     const children = React.Children.map(this.props.children, child =>
       React.cloneElement(child, {
-        identifier: that.state.identifier
+        identifier: identifier,
+        phaseId: phaseId,
+        discussionPhaseId: discussionPhaseId
```

Comment:
Could make the switch to Context API here, wouldn't that be easier? We are using React 16.3.1, so it's available.

Reply:
This could be an alternative yes. Maybe next time, if we merge the Main and App components for example.